src/components/graph/ExerciseGraphLayout.ts

```diff
@@ -0,0 +1,337 @@
+    )
+    this.conceptLayers = this.constructConceptLayers(this.exerciseLayers)
+  }
+
+  /**
+   * Use the adjacency matrix to find the in-degree of each
+   * node (the number of edges pointing to the node)
+   *
+   * Return an array of numbers, where each index (representing
+   * an ExerciseGraphNode) holds the indegree value
+   */
+  private findIndegreeOfExercises(): Array<number> {
+    const indegrees = new Array(this.adjacency.length).fill(0)
+    for (const tos of this.adjacency) {
+      for (const to of tos) {
+        indegrees[to] += 1
+      }
+    }
+    return indegrees
+  }
+
+  /**
+   * Nodes with inDegree of 0 are source nodes (roots) of the graph
+   */
+  private findSourceExercises(inDegrees: Array<number>): Array<number> {
+    const sources: Array<number> = []
+    inDegrees.forEach((inDegree, i) => {
+      if (inDegree === 0) sources.push(i)
+    })
+    return sources
+  }
+
+  /**
+   * Perform a recursive search to find the length of longest path
+   * to each exercise from any source.  This places an exercise
+   * after all of its prerequisites.
+   */
+  private findExerciseDepths(sources: Array<number>): Array<number> {
+    const maxDepths = new Array(this.adjacency.length).fill(0)
+    for (const source of sources) {
+      const traverse = (index: number, depth = 0): void => {
+        if (depth > maxDepths[index]) {
+          maxDepths[index] = depth
+        }
+
+        const children = this.adjacency[index]
+        children.forEach((child) => traverse(child, depth + 1))
+      }
+      traverse(source)
+    }
+    return maxDepths
+  }
+
+  /**
+   * order the exercises by depth then index into layers
+   */
+  private constructExerciseLayers(
+    exerciseDepths: Array<number>,
+    maxDepth: number
+  ): Array<Array<ExerciseGraphNode>> {
+    const exerciseLayers = new Array(maxDepth + 1)
+
+    exerciseDepths.forEach((nodeDepth, nodeIndex) => {
+      exerciseLayers[nodeDepth] = exerciseLayers[nodeDepth] ?? []
+      exerciseLayers[nodeDepth].push(this.graph.lookupByIndex.get(nodeIndex))
+    })
+
+    return exerciseLayers
+  }
+
+  /**
+   * order the concepts without exercises by the depth of their dependent exercise
+   */
+  private constructConceptLayers(
+    exerciseLayers: Array<Array<ExerciseGraphNode>>
+  ): Array<Array<string>> {
+    // create map for layer position for a missing concept in case the concept is listed
+    // by multiple exercises
+    const conceptsVisited = new Map<string, boolean>()
+
+    // compute missing concept layers
+    const conceptLayers = new Array<Array<string>>()
+    exerciseLayers.forEach((layer, i) => {
+      conceptLayers.push([])
+      layer.forEach((node) => {
+        const missingPrereqs = this.graph.lookupMissingConceptsForExercise.get(
+          node.slug
+        )
+        if (!missingPrereqs) return
+
+        missingPrereqs.forEach((prereq) => {
+          const visited = conceptsVisited.get(prereq)
+          if (visited) return
```

Comment:
```suggestion
          if (visited) {
            return
          }
```

BTW, is this meant to "return" from the function or stop iteration? I think it's the former, but then you might want to be a bit more explicit using `.filter`:

```typescript
missingPrereqs.filter((prereq) => !conceptsVisited.get(prereq))
              .forEach((prereq) => ... )
```